Add Shelly Pro 3 EM

wimaha opened the pull request:
- Support for Shelly Pro 3 EM
- updated Docs

wimaha committed:
Added Support for Shelly Pro 3 EM

premultiply commented:
/cc @thierolm

wimaha commented:
Can someone help me to understand how to fix the GitHub check "Error: Process completed with exit code 2."?

andig commented:
/cc @thierolm

premultiply committed:
add newline

andig commented:
Brauchen wir wirklich noch eine zweite Shellyimplementierung oder kann das nicht durch die vorhandene mit abgedeckt werden?

wimaha commented:
Die vorhandene Shellyimplementierung ist für Schalter. Ich denke es ist einfacher das Energy Meter und die Schalter getrennt zu haben.

andig commented:
Das können wir machen, aber dann brauchts eine bessere Idee wie dupliziertes Code vermieden werden kann. Hier gibts zuviele Überschneidungen.

wimaha committed:
Merge branch 'evcc-io:master' into master

wimaha committed:
Code clean

wimaha committed:
Merge branch 'master' of https://github.com/wimaha/evcc

wimaha commented:
Gut, ich ändere den Code so ab, dass die aktuellen Überschneidungen aus der bereits vorhandenen Shelly-Implementierung benutzt werden.

andig commented:
Das wär Klasse. Falls dafür was refactored werden muss kein Problem!

andig commented:
Code bitte mit `gofmt` formatieren, am Besten VSCode nehmen.

wimaha committed:
Merge branch 'evcc-io:master' into master

wimaha committed:
Refactored

I have integrated the new code into the code of the existing meter. As a result, there are no longer any duplications.

wimaha committed:
Merge branch 'master' of https://github.com/wimaha/evcc

andig commented on templates/definition/meter/shelly-pro-3em.yaml:
```diff
@@ -0,0 +1,22 @@
+  - name: user
+  - name: password
+    mask: true
+  - name: channel
```
Wird der channel denn verwendet?

wimaha replied:
Nein, werde ich entfernen.

wimaha replied:
Ist entfernt

andig commented on templates/definition/meter/shelly-pro-3em.yaml:
```diff
@@ -0,0 +1,22 @@
+    default: 192.0.2.2
+  - name: user
+  - name: password
+    mask: true
```
pwd ist immer maskiert (siehe defaults.yaml)

wimaha replied:
Ist entfernt

andig commented on templates/definition/meter/shelly-pro-3em.yaml:
```diff
@@ -0,0 +1,22 @@
+  - name: password
+    mask: true
+  - name: channel
+    default: 0
```
Null defaults sind automatisch, kann weg

wimaha replied:
ist entfernt

andig commented on meter/shelly/connection.go:
```diff
@@ -18,11 +18,12 @@ type Connection struct {
 	gen        int    // Shelly api generation
-	devicetype string // Shelly device type
+	devicetype string //Gen1 DeviceType
+	deviceType string // Shelly device type
```
Warum 2?

wimaha replied:
DeviceType nicht länger notwendig

andig commented:
Das ist schon ein super Schritt, aber immer noch nicht rund. Jetzt hat z.B. das EnergyMeter Eigenschaften, die es gar nicht besitzt wie z.B. ein- und ausschalten. Auf den ersten Blick würde es sich noch anbieten, die connection wirklich auf die Verbindung, lesen+schreiben zu reduzieren und dann `Switch` und `Meter` zusätzlich einzuführen, welche die physikalischen Geräte abstrahieren. 
Wenn Du das machst dann würde sich vmtl auch anbieten, einen extra `shelly-energy-meter.go` zu nutzen so dass der `DeviceType` vollständig entfallen könnte.

thierolm commented:
> Die vorhandene Shellyimplementierung ist für Schalter. Ich denke es ist einfacher das Energy Meter und die Schalter getrennt zu haben.

Sehe ich nicht so, meter ist meter ...

thierolm commented:
Warum wird der Device Type "hart" überschrieben ? ... Ich sehe den PR jetzt erst, gebt mir bitte die Zeit mich einzulesen ...

andig commented:
Das ist ein anderer Typ- bei Dir die Art des EM, hier die Differenzierung switch/meter

thierolm commented:
> Das ist ein anderer Typ- bei Dir die Art des EM, hier die Differenzierung switch/meter

Dann iost das semantisch aber keine DeviceType, sondern eher der DeviceMode!!!

@wimaha:  Bitte den Parameter entsprechend umbenennen und an dem bestehenden DeviceType Parameter nicht drehen!

wimaha commented:
Ich werde versuchen es so, wie von @andig beschrieben umzusetzen. Damit dürfte der Parameter dann wieder entfallen.

wimaha committed:
Merge branch 'evcc-io:master' into master

wimaha committed:
New Version

Switch and Meter were moved to seperate files with structs.

wimaha commented:
Ich habe nun ShellySwitch und ShellyMeter neu eingeführt und somit die beiden Typen sauber getrennt. Connection wird von beiden verwendet. Schaut mal drüber, ich markiere es wieder als Ready for review.

Ich habe leider kein Shelly 1 PM oder ähnlich, um den Switch-Teil zu testen. Der Meter-Teil funktioniert bei mir.

andig commented:
Gefällt mir gut! Das diff lässt sich nochmal deutlich reduzieren, das würde ich mir anschauen wenn @thierolm einverstanden ist.

andig committed:
Simplify

andig committed:
wip

thierolm commented:
Für mich fein, ich kann aber erst heute Abend testen.

andig committed:
wip

andig committed:
Simplify

andig committed:
Fix test

andig commented on templates/definition/meter/shelly-pro-3em.yaml:
```diff
@@ -0,0 +1,18 @@
+  - name: user
+  - name: password
+render: |
+  type: shelly-energy-meter
```
```suggestion
  type: shelly-energymeter
```

andig commented on meter/shelly_energymeter.go:
```diff
@@ -0,0 +1,33 @@
+
+// Shelly meter implementation
+func init() {
+	registry.Add("shelly-energy-meter", NewShellyEnergyMeterFromConfig)
```
```suggestion
	registry.Add("shelly-energymeter", NewShellyEnergyMeterFromConfig)
```

thierolm commented on templates/docs/meter/shelly-1pm_0.yaml:
```diff
@@ -11,4 +11,4 @@ render:
       host: 192.0.2.2 # IP-Adresse oder Hostname
       user: # Benutzerkonto (bspw. E-Mail Adresse, User Id, etc.) # Optional
       password: # Passwort des Benutzerkontos (bei führenden Nullen bitte in einfache Hochkommata setzen) # Optional
```
In den Templates müssen wir den Hinweis noch von einfache auf doppelte Hochkommata ändern:
`... bei führenden Nullen bitte in doppelte Hochkommata setzen ...`

wimaha committed:
Update templates/definition/meter/shelly-pro-3em.yaml

Co-authored-by: andig <cpuidle@gmx.de>

wimaha committed:
Update meter/shelly_energymeter.go

Co-authored-by: andig <cpuidle@gmx.de>

wimaha committed:
Merge branch 'evcc-io:master' into master

thierolm commented:
Ich habe alle mir zur Verfügung stehenden Shellies mit dem PR erfolgreich testen können:
```bash
thierolm@LAPTOP-KQMMVN3H:~/git-repos/evcc$ git branch -v
  master         5a964c75 Update evcc.service (#6474)
* pr/wimaha/6457 39846bfd Merge branch 'evcc-io:master' into master
thierolm@LAPTOP-KQMMVN3H:~/git-repos/evcc$ ./evcc -c ./evcc_shellytest.yaml meter
[main  ] INFO 2023/02/27 17:41:30 evcc 0.113.2 (39846bfd)
[main  ] INFO 2023/02/27 17:41:30 using config file: ./evcc_shellytest.yaml
[db    ] INFO 2023/02/27 17:41:30 using sqlite database: /home/thierolm/.evcc/evcc.db
shelly1
-------
Power:  0W
Energy: 0.0kWh

balkonpv-shelly-1pm-gen2
------------------------
Power:  1W
Energy: 11.4kWh

shelly-1pm-gen2
---------------
Power:  35W
Energy: 40.7kWh

shelly1pm
---------
Power:  0W
Energy: 107.9kWh

thierolm@LAPTOP-KQMMVN3H:~/git-repos/evcc$ ./evcc -c ./evcc_shellytest.yaml charger
[main  ] INFO 2023/02/27 17:41:36 evcc 0.113.2 (39846bfd)
[main  ] INFO 2023/02/27 17:41:36 using config file: ./evcc_shellytest.yaml
[db    ] INFO 2023/02/27 17:41:36 using sqlite database: /home/thierolm/.evcc/evcc.db
Power:         35W
Energy:        40.7kWh
Charge status: C
Enabled:       true
```

ecfmau commented:
Hallo, es scheint, dass die Berechnung der geladenen Energie um einen Faktor 1000 falsch ist, siehe #7080.
